ledger/streaming/client/src/block_validator.rs

```diff
@@ -0,0 +1,261 @@
+                        if let Some(ledger) = ledger {
+                            for key_image in &block_contents.key_images {
+                                // Check if the key image is already in the local ledger.
+                                match ledger.contains_key_image(key_image) {
+                                    Ok(contains_key_image) => {
+                                        if contains_key_image
+                                            || additional_key_images.contains(key_image)
+                                        {
+                                            return future::ready(Some(Err(
+                                                StreamError::BlockValidation(
+                                                    "Contains spent key image".to_string(),
+                                                ),
+                                            )));
+                                        }
+                                    }
+                                    Err(err) => {
+                                        return future::ready(Some(Err(StreamError::DBAccess(
+                                            err.to_string(),
+                                        ))));
+                                    }
+                                }
+                                additional_key_images.insert(*key_image);
+                            }
+                        }
+
+                        // Compute the hash of the block
+                        let derived_block_id = compute_block_id(
+                            block.version,
+                            &block.parent_id,
+                            block.index,
+                            block.cumulative_txo_count,
+                            &block.root_element,
+                            &block_contents.hash(),
+                        );
+
+                        // The block's ID must agree with the merkle hash of its transactions.
+                        if block.id != derived_block_id {
+                            return future::ready(Some(Err(StreamError::BlockValidation(
+                                "Hash of transactions don't match claimed block id".to_string(),
+                            ))));
+                        }
+
+                        *prev_block_id = block.id.clone();
+                        future::ready(Some(Ok(component)))
+                    }
+                    Err(err) => future::ready(Some(Err(err))),
+                }
+            },
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use mc_common::logger::{log, test_with_logger, Logger};
+    use mc_ledger_db::test_utils::{get_mock_ledger, MockLedger};
+    use mc_ledger_streaming_api::test_utils::{
+        make_components, make_quorum_set, mock_stream_from_components, stream,
+        stream::SimpleMockStream,
+    };
+    use mc_transaction_core::BlockIndex;
+
+    use super::*;
```

Comment:
Remove extra blank line and sort
```suggestion
    use super::*;
    use mc_common::logger::{log, test_with_logger, Logger};
    use mc_ledger_db::test_utils::{get_mock_ledger, MockLedger};
    use mc_ledger_streaming_api::test_utils::{
        make_components, make_quorum_set, mock_stream_from_components, stream,
        stream::SimpleMockStream,
    };
    use mc_transaction_core::BlockIndex;
```

Reply:
Addressed by rebase in separate branch